instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressionStrategy.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.context.Context;
+import java.util.Arrays;
+import java.util.List;
+
+abstract class SpanSuppressionStrategy {
+
+  private static final SpanSuppressionStrategy SERVER_STRATEGY = new SuppressIfSameType(
+      Arrays.asList(SpanKey.SERVER));
+  private static final SpanSuppressionStrategy CONSUMER_STRATEGY = new SuppressIfSameType(
+      Arrays.asList(SpanKey.CONSUMER));
+
+  static SpanSuppressionStrategy from(List<SpanKey> clientSpanKeys) {
+    if (clientSpanKeys.isEmpty()) {
+      return new CompositeStrategy(
+          NeverSuppress.INSTANCE,
+          SERVER_STRATEGY,
+          CONSUMER_STRATEGY);
+    }
+    return new CompositeStrategy(
+        new SuppressIfSameType(clientSpanKeys),
+        SERVER_STRATEGY,
+        CONSUMER_STRATEGY);
+  }
+
+  abstract Context storeInContext(SpanKind spanKind, Context context, Span span);
+
+  abstract boolean shouldSuppress(SpanKind spanKind, Context parentContext);
+
+  static final class SuppressIfSameType extends SpanSuppressionStrategy {
+
+    private final List<SpanKey> outgoingSpanKeys;
+
+    SuppressIfSameType(List<SpanKey> outgoingSpanKeys) {
+      this.outgoingSpanKeys = outgoingSpanKeys;
+    }
+
+    @Override
+    Context storeInContext(SpanKind spanKind, Context context, Span span) {
+      for (SpanKey outgoingSpanKey : outgoingSpanKeys) {
+        context = outgoingSpanKey.with(context, span);
+      }
+      return context;
+    }
+
+    @Override
+    boolean shouldSuppress(SpanKind spanKind, Context parentContext) {
+        for (SpanKey outgoingSpanKey : outgoingSpanKeys) {
+          if (outgoingSpanKey.fromContextOrNull(parentContext) == null) {
+            return false;
+          }
+        }
+        return true;
```

Comment:
There's a bit too much indentation here 🦅 👁️ 

```suggestion
      for (SpanKey outgoingSpanKey : outgoingSpanKeys) {
        if (outgoingSpanKey.fromContextOrNull(parentContext) == null) {
          return false;
        }
      }
      return true;
```